src/pool/pool_tbb.c

```diff
@@ -0,0 +1,200 @@
+
+struct mem_pool_policy_s {
+    raw_alloc_tbb_type pAlloc;
+    raw_free_tbb_type pFree;
+    size_t granularity;
+    int version;
+    unsigned fixed_pool : 1, keep_all_memory : 1, reserved : 30;
+};
+
+struct tbb_callbacks {
+    void *(*pool_malloc)(void *, size_t);
+    void *(*pool_realloc)(void *, void *, size_t);
+    void *(*pool_aligned_malloc)(void *, size_t, size_t);
+    bool (*pool_free)(void *, void *);
+    int (*pool_create_v1)(intptr_t, const struct mem_pool_policy_s *, void **);
+    bool (*pool_destroy)(void *);
+    void *(*pool_identify)(void *object);
+    size_t (*pool_msize)(void *, void *);
+};
+
+struct tbb_memory_pool {
+    umf_memory_provider_handle_t mem_provider;
+    void *tbb_pool;
+};
+
+static struct tbb_callbacks g_tbb_ops;
+static pthread_once_t tbb_is_initialized = PTHREAD_ONCE_INIT;
+
+void load_tbb_symbols(void) {
+    fprintf(stdout, "Initializing TBB symbols\n");
+    const char so_name[] = "libtbbmalloc.so.2";
+    void *tbb_handle = dlopen(so_name, RTLD_LAZY);
+    if (!tbb_handle) {
+        fprintf(stderr, "%s not found.\n", so_name);
+        abort();
+    }
+
+    struct tbb_callbacks tbb_ops;
+
+    *(void **)&tbb_ops.pool_malloc =
+        dlsym(tbb_handle, "_ZN3rml11pool_mallocEPNS_10MemoryPoolEm");
+    *(void **)&tbb_ops.pool_realloc =
+        dlsym(tbb_handle, "_ZN3rml12pool_reallocEPNS_10MemoryPoolEPvm");
+    *(void **)&tbb_ops.pool_aligned_malloc =
+        dlsym(tbb_handle, "_ZN3rml19pool_aligned_mallocEPNS_10MemoryPoolEmm");
+    *(void **)&tbb_ops.pool_free =
+        dlsym(tbb_handle, "_ZN3rml9pool_freeEPNS_10MemoryPoolEPv");
+    *(void **)&tbb_ops.pool_create_v1 = dlsym(
+        tbb_handle,
+        "_ZN3rml14pool_create_v1ElPKNS_13MemPoolPolicyEPPNS_10MemoryPoolE");
+    *(void **)&tbb_ops.pool_destroy =
+        dlsym(tbb_handle, "_ZN3rml12pool_destroyEPNS_10MemoryPoolE");
+    *(void **)&tbb_ops.pool_identify =
+        dlsym(tbb_handle, "_ZN3rml13pool_identifyEPv");
+    *(void **)&tbb_ops.pool_msize =
+        dlsym(tbb_handle, "_ZN3rml10pool_msizeEPNS_10MemoryPoolEPv");
+
+    if (!tbb_ops.pool_malloc || !tbb_ops.pool_realloc ||
+        !tbb_ops.pool_aligned_malloc || !tbb_ops.pool_free ||
+        !tbb_ops.pool_create_v1 || !tbb_ops.pool_destroy ||
+        !tbb_ops.pool_identify) {
+        fprintf(stderr, "Could not find symbols in %s.\n", so_name);
+        dlclose(tbb_handle);
+        abort();
+    }
+
+    g_tbb_ops = tbb_ops;
+}
+
+void *tbb_raw_alloc_wrapper(intptr_t pool_id, size_t *raw_bytes) {
+    void *resPtr;
+    struct tbb_memory_pool *pool = (struct tbb_memory_pool *)pool_id;
+    enum umf_result_t ret =
+        umfMemoryProviderAlloc(pool->mem_provider, *raw_bytes, 0, &resPtr);
+
+    if (ret != UMF_RESULT_SUCCESS) {
+        return NULL;
+    }
+
+    return resPtr;
+}
+
+void tbb_raw_free_wrapper(intptr_t pool_id, void *ptr, size_t bytes) {
+    struct tbb_memory_pool *pool = (struct tbb_memory_pool *)pool_id;
+    enum umf_result_t ret =
```

Comment:
Probably we should store the error in TLS and return it from tbb_free if necessary. Or perhaps we could just store it in the tbb_memory_pool struct?

Reply:
`get_last_allocation_error()` retrieves `umf_result_t` representing the error of the last failed **allocation operation** in this thread (**malloc, calloc, realloc, aligned_malloc**) - not free.

Reply:
That's true, but we can return the result directly from umfPoolFree (`tbb_free`). Right now we always return SUCCESS from what I see.

Reply:
In `tbb_free()` yes, we can, but a return value of `pool_free()` is never used or checked in the TBB repo ... Done anyway.